core/templates/services/audio-player.service.spec.ts

```diff
@@ -0,0 +1,475 @@
+    beforeEach(() => {
+      spyOn(howler, 'Howl').and.returnValue({
+        on: (evt, func) => {
+          if (evt === 'load') {
+            func();
+          }
+        },
+        stop: () => {
+          console.error('Howl.stop');
+        },
+        pause: () => {
+          console.error('Howl.pause');
+        },
+        playing: () => {
+          return false;
+        },
+        play: () => {
+          console.error('Howl.play');
+          return 2;
+        },
+        seek: (num?: number) => {
+          if (typeof num !== 'undefined') {
+            console.error('Howl.seek ', num);
+            return num;
+          }
+          return 10;
+        },
+        duration: () => {
+          return 30;
+        }
+      } as unknown as Howl);
+      spyOn(assetsBackendApiService, 'loadAudio').and.returnValue(
+        Promise.resolve({
+          data: new Blob(),
+          filename: 'test.mp3'
+        }));
+    });
+
+    it('should load when user plays audio', fakeAsync(async() => {
+      audioPlayerService.loadAsync('test.mp3').then(
+        successHandler, failHandler);
+      flushMicrotasks();
+
+      expect(successHandler).toHaveBeenCalled();
+      expect(failHandler).not.toHaveBeenCalled();
+    }));
+
+    it('should not load track when a track has laready been loaded',
+      fakeAsync(() => {
+        audioPlayerService.loadAsync('test.mp3');
+        flushMicrotasks();
+        audioPlayerService.loadAsync('test.mp3');
+        flushMicrotasks();
+
+        expect(assetsBackendApiService.loadAudio).toHaveBeenCalledTimes(1);
+      }));
+
+    it('should play when user clicks the play button', fakeAsync(() => {
+      spyOn(console, 'error');
+
+      audioPlayerService.loadAsync('test.mp3');
+      flushMicrotasks();
+
+      audioPlayerService.play();
+      tick(501);
+      discardPeriodicTasks();
+
+      // The play function of the Howl calss is called when the user clicks
+      // the  play button. Since the Howl class is mocked, a cosole.error stmt
+      // is placed inside the play function and is tested if the console.error
+      // stmt inside it triggered.
+      expect(console.error).toHaveBeenCalledWith('Howl.play');
+    }));
+
+    it('should not play when audio is already playing', fakeAsync(() => {
+      spyOn(console, 'error');
+      spyOn(audioPlayerService, 'isPlaying').and.returnValue(true);
+
+      audioPlayerService.loadAsync('test.mp3');
+      flushMicrotasks();
+
+      audioPlayerService.play();
+      discardPeriodicTasks();
+
+      // The play function of the Howl calss is called when the user clicks
+      // the  play button. Since the Howl class is mocked, a cosole.error stmt
+      // is placed inside the play function and is tested if the console.error
+      // stmt inside it triggered.
+      expect(console.error).not.toHaveBeenCalledWith('Howl.play');
+    }));
+
+    it('should not pause when audio is not playing', fakeAsync(() => {
```

Comment:
Can we make descriptions more clear here and elsewhere?

Reply:
done